src/sqlfluff/core/parser/segments/raw.py

```diff
@@ -148,3 +151,51 @@ def get_end_pos_marker(self):
     def get_start_pos_marker(self):
         """Return the pos marker at the start of this segment."""
         return self.pos_marker
+
+
+class RawSegmentFactory:
+    optional = False
+    _anti_template = None
+
+    def __init__(self, cls, classname, _template, _name, **kwargs):
+        # if classname.startswith("semicolon"):
+        #     import pdb; pdb.set_trace()
+        self.cls = cls
+        self.__name__ = classname
```

Comment:
Do we need to play with `__name__`? This feels like overcomplication. If it's to replicate some of the features of `RawSegment` then I wonder if we should get rid of those features.

Reply:
Eventually, perhaps not. When I work on things like this, I tend to take an exploratory approach and try to get various tests working, then go back and clean up the ugly bits. This is definitely ugly.

I was thinking last night, I think we're implicitly moving in a direction where we don't use `@classmethod` anymore, nor do we use `type()`, `isinstance()`, `issubclass()`, etc.

From my previous experience with parsers in C++, it's common to have parallel class hierarchies, basically descriptors and then parser objects which are "instances" (but not in the C++ sense) of those descriptors. Essentially, we build our own type system in the parser.

I wonder if we have to do 100% of this work now (or ever). I've had some success thus far by adding the `factory` property to `ParseContext`. We can see how far that takes us.